readme.md
Add Plugin Directories to WordPress
===================================

A work in progress related to a [WordPress stack exchange question](http://wordpress.stackexchange.com/questions/43262/add-multiple-plugin-directories).


EXAMPLE

For an example, take a look at the `example_plugin` folder and the `register_addt_plugins_dir.php` file.

You need to copy this folder to your actual plugins folder and then activate it from within your plugins list. It only works if the main plugin is activated.

*) the one you defined with the `WP_PLUGIN_DIR` or `WPMU_PLUGIN_DIR` in your wp-config.php file - or the default `plugins` folder in your install.

changelog.txt
*0.1*
Initial version.
*0.2*
Clean Up & code styling alignment.
*0.3*
Minor styling fixes.
*0.4*
Moved to OOP concept.
*0.5*
Improved API: now supports different plugin locations aside from the `WP_CONTENT_DIR`.
*0.5.1*
Minor fix for left over debug code.
*0.5.2*
JS styling for readability.
*0.6*
Removed "activate" link when plugin is already active, as suggested by Julien Chaumond in Issue #3.
*0.7*
Added the "adp_root_{$root}" filter to allow completely custom locations. Moved changelog to separate file.
*0.7.1*
Fix script path for remote locations.
*0.7.2*
Renamed main plugin file to index.php to avoid accidental directory listing.
*0.7.3*
Reverted 0.7.2 to avoid plugin deactivation. Added index.html file instead.
*0.8*
Moved Bootstrap to OOP construct. Regular child plugins are now allowed to use the default priority.
*0.9*
Added better update message (as seen here).


inc/core.php
<?php
! defined( 'ABSPATH' ) AND exit();



if ( ! class_exists( 'CD_APD_Core' ) )
{

/**
 * Core
 * 
 * @author     Christopher Davis, Franz Josef Kaiser
 * @license    GPL
 * @copyright  © Christopher Davis, Franz Josef Kaiser 2011-2012
 * 
 * @package    WordPress
 * @subpackage Additional Plugin Directories: Core
 */
class CD_APD_Core
{
	/**
	 * 
	 */
	public function __construct()
	{
		add_action( 'plugins_loaded', array( $this, 'load_plugins' ), 99 );
	}


	/**
	 * Loads additional plugins from custom directories. 
	 * To add a directory, you must do so in a plugin (hooked into `plugins_loaded` with a low priority).
	 * 
	 * @since  0.1
	 * @return void
	 */
	public function load_plugins()
	{
		global $wp_plugin_directories;
	
		empty( $wp_plugin_directories ) AND $wp_plugin_directories = array();
	
		foreach ( array_keys( $wp_plugin_directories ) as $key )
		{
			$active = get_option( "active_plugins_{$key}", array() );
			foreach( $active as $a )
			{
				if ( file_exists( "{$wp_plugin_directories[ $key ]['dir']}/{$a}" ) )
				{
					include_once( "{$wp_plugin_directories[ $key ]['dir']}/{$a}" );
				}
			}
		}
	}


	/**
	 * Get the valid plugins from the custom directory
	 * 
	 * @since  0.1
	 * @param  $dir_key The `key` of our custom plugin directory
	 * @return array A list of the plugins
	 */
	public function get_plugins_from_cache( $dir_key ) 
	{
		global $wp_plugin_directories;

		// invalid dir key? bail
		if ( ! isset( $wp_plugin_directories[ $dir_key ] ) )
			return array();

		$plugin_root = $wp_plugin_directories[ $dir_key ]['dir'];

		if ( ! $cache_plugins = wp_cache_get( 'plugins', 'plugins') )
			$cache_plugins = array();

		if ( isset( $cache_plugins[ $dir_key ] ) )
			return $cache_plugins[ $dir_key ];

		$wp_plugins = array();

		$plugins_dir = @ opendir( $plugin_root );
		$plugin_files = array();
		if ( $plugins_dir )
		{
			while ( false !== ( $file = readdir( $plugins_dir ) ) )
			{
				if ( '.' === substr( $file, 0, 1 ) )
					continue;

				if ( is_dir( "{$plugin_root}/{$file}" ) )
				{
					$plugins_subdir = @ opendir( "{$plugin_root}/{$file}" );
					if ( $plugins_subdir )
					{
						while ( ( $subfile = readdir( $plugins_subdir ) ) !== false )
						{
							if ( '.' === substr( $subfile, 0, 1 ) )
								continue;

							if ( '.php' === substr( $subfile, -4 ) )
								$plugin_files[] = "{$file}/{$subfile}";
						}
						closedir( $plugins_subdir );
					}
				}
				else
				{
					'.php' === substr( $file, -4 ) AND $plugin_files[] = $file;
				}
			}
			closedir( $plugins_dir );
		}

		if ( empty( $plugin_files ) )
			return $wp_plugins;

		foreach ( $plugin_files as $plugin_file )
		{
			if ( ! is_readable( "{$plugin_root}/{$plugin_file}" ) )
				continue;

			// Do not apply markup/translate as it'll be cached.
			$plugin_data = get_plugin_data( "{$plugin_root}/{$plugin_file}", false, false );

			if ( empty ( $plugin_data['Name'] ) )
				continue;

			$wp_plugins[ trim( $plugin_file ) ] = $plugin_data;
		}

		uasort( $wp_plugins, '_sort_uname_callback' );

		// Setup cache, if we ain't already got one.
		// If we got one, we already returned the cached plugins.
		$cache_plugins[ $dir_key ] = $wp_plugins;
		wp_cache_set( 'plugins', $cache_plugins, 'plugins' );

		return $wp_plugins;
	}


	/**
	 * Custom plugin activation function.
	 * 
	 * @since  0.1
	 * @return void
	 */
	public function activate_plugin( $plugin, $context, $silent = false ) 
	{
		$plugin = trim( $plugin );

		$redirect = add_query_arg( 'plugin_status', $context, admin_url( 'plugins.php' ) );
		$redirect = apply_filters( 'custom_plugin_redirect', $redirect );

		$current = get_option( "active_plugins_{$context}", array() );

		$valid = $this->validate_plugin( $plugin, $context );

		if ( is_wp_error( $valid ) )
			return $valid;

		if ( ! in_array( $plugin, $current ) )
		{
			if ( ! empty( $redirect ) )
			{
				// we'll override this later if the plugin can be included without fatal error
				wp_redirect( add_query_arg( 
					 '_error_nonce'
					,wp_create_nonce( "plugin-activation-error_{$plugin}" )
					,$redirect 
				) );
			}

			ob_start();
			include_once( $valid );

			if ( ! $silent )
			{
				do_action( 'custom_activate_plugin', $plugin, $context );
				do_action( "custom_activate_{$plugin}", $context );
			}

			$current[] = $plugin;
			sort( $current );
			update_option( "active_plugins_{$context}", $current );

			if ( ! $silent )
			{
				do_action( 'custom_activated_plugin', $plugin, $context );
			}

			if ( ob_get_length() > 0 )
			{
				$output = ob_get_clean();
				return new WP_Error( 'unexpected_output', __( 'The plugin generated unexpected output.' ), $output );
			}
			ob_end_clean();
		}

		return true;
	}


	/**
	 * Deactivate custom plugins
	 * 
	 * @since  0.1
	 * @return void
	 */
	public function deactivate_plugins( $plugins, $context, $silent = false )
	{
		$current = get_option( "active_plugins_{$context}", array() );

		foreach ( (array) $plugins as $plugin ) 
		{
			$plugin = trim( $plugin );
			if ( ! in_array( $plugin, $current ) ) continue;

			if ( ! $silent )
				do_action( 'custom_deactivate_plugin', $plugin, $context );

			$key = array_search( $plugin, $current );
			if ( false !== $key )
			{
				array_splice( $current, $key, 1 );
			}

			if ( ! $silent )
			{
				do_action( "custom_deactivate_{$plugin}", $context );
				do_action( 'custom_deactivated_plugin', $plugin, $context );
			}
		}

		update_option( "active_plugins_{$context}", $current );
	}


	/**
	 * Checks to see whether the plugin and is valid and can be activated.
	 * 
	 * @uses validate_file To make sure the plugin name is okay.
	 * @param  string $plugin
	 * @return array $context
	 * @return WP_Error|string WP_Error object on failure, the plugin to include on success.
	 */
	public function validate_plugin( $plugin, $context ) 
	{
		$rv = true;
		if ( validate_file( $plugin ) )
		{
			$rv = new WP_Error( 'plugin_invalid', __( 'Invalid plugin path.' ) );
		}

		global $wp_plugin_directories;

		if ( ! isset( $wp_plugin_directories[ $context ] ) )
		{
			$rv = new WP_Error( 'invalid_context', __( 'The context for this plugin does not exist' ) );
		}
	
		$dir = $wp_plugin_directories[ $context ]['dir'];
		if ( ! file_exists( "{$dir}/{$plugin}" ) )
		{
			$rv = new WP_Error( 'plugin_not_found', __( 'Plugin file does not exist.' ) );
		}

		$installed_plugins = $this->get_plugins_from_cache( $context );
		if ( ! isset( $installed_plugins[ $plugin ] ) )
		{
			$rv = new WP_Error( 'no_plugin_header', __('The plugin does not have a valid header.') );
		}

		$rv = "{$dir}/{$plugin}";

		return $rv;
	}
} // END Class CD_APD_Core

} // endif;

inc/admin.php
<?php
! defined( 'ABSPATH' ) AND exit();



if ( ! class_exists( 'CD_APD_Admin' ) )
{

/**
 * Admin/Factory
 * 
 * @author     Christopher Davis, Franz Josef Kaiser
 * @license    GPL
 * @copyright  © Christopher Davis, Franz Josef Kaiser 2011-2012
 * 
 * @package    WordPress
 * @subpackage Additional Plugin Directories: Admin/Factory
 */
class CD_APD_Admin extends CD_APD_Core
{
	/**
	 * Instance
	 * 
	 * @since  0.3
	 * @access protected
	 * @var    object
	 */
	protected static $instance;


	/**
	 * The container for all of our custom plugins
	 * 
	 * @since  0.1
	 * @access protected
	 * @var    array
	 */
	protected $plugins = array();


	/**
	 * What custom actions are we allowed to handle here?
	 * 
	 * @since  0.1
	 * @access protected
	 * @var    array
	 */
	protected $actions = array();


	/**
	 * The original count of the plugins
	 * 
	 * @since  0.1
	 * @access protected
	 * @var    int
	 */
	protected $all_count = 0;


	/**
	 * Creates a new static instance
	 * 
	 * @since  0.3
	 * @static
	 * @return void
	 */
	public static function instance()
	{
		null === self :: $instance AND self :: $instance = new self;
		return self :: $instance;
	}


	/**
	 * constructor
	 * 
	 * @since  0.1
	 * @return void
	 */
	public function __construct()
	{
		add_action( 'plugins_loaded', array( $this, 'setup_actions' ), 11 );
		add_action( 'load-plugins.php', array( $this, 'init' ) );
	}


	/**
	 * Sets up which actions we can handle with this plugin. We'll use this
	 * to catch activations and deactivations as the normal way won't work.
	 * Has the filter 'custom_plugin_actions' to allow extensions.
	 * 
	 * @since  0.1
	 * @return void
	 */
	public function setup_actions()
	{
		$this->actions = apply_filters( 
			'custom_plugin_actions',
			array(
				'custom_activate',
				'custom_deactivate'
			)
		);
	}


	/**
	 * Makes the magic happen. Loads all the other hooks to modify the plugin list table
	 * 
	 * @since  0.1
	 * @return void
	 */
	public function init()
	{
		global $wp_plugin_directories;

		$screen = get_current_screen();

		$this->get_plugins();

		$this->handle_actions();

		add_filter( "views_{$screen->id}", array( $this, 'views' ) );

		// check to see if we're using one of our custom directories
		if ( $this->get_plugin_status() )
		{
			// Disable default "advanced" plugins. Inside the callback, 
			// the 2nd arg is either for "Must Use" => 'mustuse' and for "DropIns" => 'dropins'.
			add_filter( 'show_advanced_plugins', '__return_false', 10, 2 );

			#add_filter( "views_{$screen->id}", array( $this, 'views_again' ) );
			add_filter( 'all_plugins', array( $this, 'filter_plugins' ) );
			// @TODO: support bulk actions
			add_filter( "bulk_actions-{$screen->id}", '__return_empty_array' );
			add_filter( 'plugin_action_links', array( $this, 'action_links' ), 10, 2 );
			add_action( 'admin_enqueue_scripts', array( $this, 'scripts' ) );
		}
	}


	/**
	 * Adds our custom plugin directories to the list of plugin types
	 *
	 * @since  0.1
	 * @param  array $views
	 * @return array $views
	 */
	public function views( $views )
	{
		global $wp_plugin_directories, $totals;

		// bail if we don't have any extra dirs
		if ( empty( $wp_plugin_directories ) ) 
			return $views;
echo '<pre>'; print_r( $totals ); echo '</pre>';
		// Add our directories to the action links
		foreach ( $wp_plugin_directories as $key => $info )
		{
			$count = count( $this->plugins[ $key ] );

			if ( ! $count ) 
				continue;

			$views[ $key ] = sprintf( 
				'<a href="%s"%s>%s <span class="count">(%d)</span></a>',
				add_query_arg( 'plugin_status', $key, 'plugins.php' ),
				$this->get_plugin_status() == $key ? ' class="current" ' : '',
				esc_html( $info['label'] ),
				$count
			);
		}

		return $views;
	}


	/**
	 * Unset inactive plugin link as it doesn't really work for this view
	 * 
	 * @since  0.1
	 * @param  array $views
	 * @return array $views
	 */
	public function views_again( $views )
	{
		if ( isset( $views['inactive'] ) ) 
			unset( $views['inactive'] );
echo '<pre> '.__LINE__.__FILE__.' '; print_r( $views ); echo '</pre>';
		return $views;
	}


	/**
	 * Filters the plugins list to include all the plugins in our custom directory
	 * 
	 * @since  0.1
	 * @param  array $plugins
	 * @return array $plugins
	 */
	public function filter_plugins( $plugins )
	{
		$key = $this->get_plugin_status();
		if ( $key )
		{
			$this->all_count = count( $plugins );
			$plugins = $this->plugins[ $key] ;
		}

		return $plugins;
	}


	/**
	 * Correct some action links so we can actually "activate" plugins
	 * 
	 * @since  0.1
	 * @param  array  $links
	 * @param  string $plugin_file
	 * @return array  $links
	 */
	public function action_links( $links, $plugin_file )
	{
		$context = $this->get_plugin_status();

		$active = get_option( "active_plugins_{$context}", array() );

		// let's just start over
		$links = array();
		if ( ! in_array( $plugin_file, $active ) )
		{
			$links['activate'] = sprintf(
				'<a href="%s" title="Activate this plugin">%s</a>',
				wp_nonce_url( 
					"plugins.php?action=custom_activate&amp;plugin={$plugin_file}&amp;plugin_status=".esc_attr( $context ), 
					"custom_activate-{$plugin_file}" 
				),
				__( 'Activate' )
			);
		}

		if ( in_array( $plugin_file, $active ) )
		{
			$links['deactivate'] = sprintf(
				'<a href="%s" title="Deactivate this plugin" class="cd-apd-deactivate">%s</a>',
				wp_nonce_url( 
					"plugins.php?action=custom_deactivate&amp;plugin={$plugin_file}&amp;plugin_status=".esc_attr( $context ), 
					"custom_deactivate-{$plugin_file}" 
				),
				__( 'Deactivate' )
			);
		}

		return $links;
	}


	/**
	 * Enqueues on JS file for fun hacks.
	 * 
	 * @uses filemtime() to set the version number of files 
	 * to their last changed date to prevent caching.
	 * 
	 * @since  0.1
	 * @uses   wp_enqueue_script()
	 * @return void
	 */
	public function scripts( $screen )
	{
		if ( 'plugins.php' !== $screen )
			return;

		wp_enqueue_script(
			 'cd-apd-js'
			,$this->scripts_file_cb( 'url' )."apd.js"
			,array( 'jquery' )
			,filemtime( $this->scripts_file_cb( 'path' )."apd.js" )
		);
		wp_localize_script(
			'cd-apd-js',
			'cd_apd',
			array(
				'count' => esc_js( $this->all_count )
			)
		);
	}


	/**
	 * Callback to get the Path or URl to register scripts.
	 * 
	 * @since  0.7.3
	 * @param  string $case (Valid are:) 'path', 'url'
	 * @param  string $sub_dir Defaults to: 'js'
	 * @return string
	 */
	public function scripts_file_cb( $case, $sub_dir = 'js' )
	{
		$root = 'path' === $case ? plugin_dir_path( __FILE__ ) : plugin_dir_url( __FILE__ );

		return substr_replace( 
			 $root
			,$sub_dir
			,strrpos( $root, basename( $root ) )
			,strlen( basename( $root ) ) 
		);
	}


	/**
	 * Fetch all the custom plugins we have!
	 * 
	 * @since  0.1
	 * @uses   get_plugins_from_cache() To fetch all the custom plugins
	 * @return void
	 */
	public function get_plugins()
	{
		global $wp_plugin_directories;

		empty( $wp_plugin_directories ) AND $wp_plugin_directories = array();

		foreach ( array_keys( $wp_plugin_directories ) as $key )
		{
			$this->plugins[ $key ] = $this->get_plugins_from_cache( $key );
		}
	}


	/**
	 * Handle activations and deactivations as the standard way will
	 * fail with "plugin file does not exist
	 *
	 * @since  0.1
	 * @return void
	 */
	public function handle_actions()
	{
		$action = isset( $_REQUEST['action'] ) ? $_REQUEST['action'] : '';

		// not allowed to handle this action? bail.
		if ( ! in_array( $action, $this->actions ) ) return;

		// Get the plugin we're going to activate
		$plugin = isset( $_REQUEST['plugin'] ) ? $_REQUEST['plugin'] : false;

		if ( ! $plugin )
			return;

		$context = $this->get_plugin_status();

		switch( $action )
		{
			case 'custom_activate':
				if ( ! current_user_can('activate_plugins') )
					wp_die( __( 'You do not have sufficient permissions to manage plugins for this site.' ) );

				check_admin_referer( "custom_activate-{$plugin}" );

				$result = $this->activate_plugin( $plugin, $context );
				if ( is_wp_error( $result ) ) 
				{
					if ( 'unexpected_output' == $result->get_error_code() ) 
					{
						wp_redirect( add_query_arg( 
							'_error_nonce',
							wp_create_nonce( "plugin-activation-error_{$plugin}" ),
							add_query_arg( 
								'plugin_status', 
								$context, 
								self_admin_url( 'plugins.php' ) 
							) 
						) );
						exit();
					}
					else 
					{
						wp_die( $result );
					}
				}

				wp_redirect( add_query_arg( 
					array( 'plugin_status' => $context, 'activate' => 'true' ), 
					self_admin_url( 'plugins.php' ) 
				) );
				exit();
				break;

			case 'custom_deactivate':
				if ( ! current_user_can( 'activate_plugins' ) )
					wp_die( __( 'You do not have sufficient permissions to deactivate plugins for this site.' ) );

				check_admin_referer( "custom_deactivate-{$plugin}" );

				$this->deactivate_plugins( $plugin, $context );

				if ( headers_sent() )
				{
					printf( 
						"<meta http-equiv='refresh' content='%s' />",
						esc_attr( "0;url=plugins.php?deactivate=true&plugin_status={$status}&paged={$page}&s={$s}" )
					);
				}
				else
				{
					wp_redirect( self_admin_url( "plugins.php?deactivate=true&plugin_status={$context}" ) );
					exit();
				}
				break;

			default :
				do_action( "custom_plugin_dir_{$action}" );
				break;
		}
	}


	/**
	 * Utility function to get the current `plugin_status`. 
	 * The key returns FALSE if our key isn't in the the custom directories
	 * 
	 * @since  0.1
	 * @return bool|string $rv False on failure, the `$wp_plugin_directories` key on success
	 */
	public function get_plugin_status()
	{
		global $wp_plugin_directories;

		$rv = false;

		if ( 
			isset( $_GET['plugin_status'] ) 
			AND in_array( $_GET['plugin_status'], array_keys( $wp_plugin_directories ) ) 
			)
		{
			$rv = $_GET['plugin_status'];
		}

		return $rv;
	}
} // END Class CD_APD_Admin

} // endif;


plugin-dirs.php
<?php
! defined( 'ABSPATH' ) AND exit();
/*
Plugin Name:  Additional Plugin Directories
Plugin URI:   http://github.com/chrisguitarguy
Description:  A framework to allow adding additional plugin directories to WordPress
Version:      1.1
Author:       Christopher Davis
Contributors: Franz Josef Kaiser, Julien Chaumond
Author URI:   http://christopherdavis.me
License:      GNU GPL 2
*/



// Avoid loading twice
if ( ! class_exists( 'dmb_bootstrap' ) )
{
	add_action( 'plugins_loaded', array( 'CD_APD_Bootstrap', 'init' ), 5 );

/**
 * Bootstrap for delayed Meta Boxes
 * 
 * @author     Franz Josef Kaiser, Christopher Davis
 * @license    GNU GPL 2
 * @copyright  © Franz Josef Kaiser, Christopher Davis 2011-2012
 * 
 * @package    WordPress
 * @subpackage Additional Plugin Directories: Bootstrap
 */
class CD_APD_Bootstrap
{
	/**
	 * Instance
	 * 
	 * @access protected
	 * @var object
	 */
	static protected $instance;


	/**
	 * The files that need to get included
	 * 
	 * @since     0.8
	 * @access    public
	 * @static
	 * @var array string Class Name w/o prefix (Hint: Naming convention!) Use the value to define if need to hook the class.
	 */
	static public $includes = array(
		 'api'   => false
		,'core'  => false
		,'admin' => true
	);


	/**
	 * Used for update notices
	 * Fetches the readme file from the official plugin repo trunk.
	 * Adds to the "in_plugin_update_message-$file" hook
	 * 
	 * @var (string)
	 */
	public $remote_changelog = 'https://raw.github.com/chrisguitarguy/WP-Plugin-Directories/master/changelog.html';


	/**
	 * Creates a new static instance
	 * 
	 * @since  0.8
	 * @static
	 * @return void
	 */
	static public function init()
	{
		null === self :: $instance AND self :: $instance = new self;
		return self :: $instance;
	}


	/**
	 * Constructor
	 * 
	 * @since  0.8
	 * @access public
	 * @return void
	 */
	public function __construct()
	{
		// Localize
		load_theme_textdomain( 'cd_apd_textdomain', plugin_dir_path( __FILE__ )."lang" );

		// Load at the end of /wp-admin/admin.php
		foreach ( self :: $includes as $inc => $init )
		{
			// Load file: trailingslashed by core
			# Tested: calling plugin_dir_path() directly saves 1/2 time 
			# instead of saving the plugin_dir_path() in a $var and recalling here
			require_once plugin_dir_path( __FILE__ )."inc/{$inc}.php";

			if ( ! $init )
				continue;

			// Build class name
			$class = "CD_APD_".ucwords( $inc );

			class_exists( $class ) AND add_action( 'plugins_loaded', array( $class, 'instance' ) );
		}

		if ( ! is_admin() )
			return;

		// Updates from GitHub
		// $ git submodule add git://github.com/franz-josef-kaiser/WordPress-GitHub-Plugin-Updater inc/updater
		add_action( 'init', array( $this, 'update_from_github' ), 0 );
	}


	/**
	 * 
	 * @since 1.0
	 * 
	 * @return void
	 */
	public function update_from_github()
	{
		global $wp_version;

		// Load the updater
		include_once plugin_dir_path( __FILE__ ).'inc/updater/updater.php';

		// Fix this strange WP bug(?)
		add_action( 'http_request_args', array( $this, 'update_request_args' ), 0, 2 );

		$host	= 'github.com';
		$http	= 'https://';
		$name	= 'chrisguitarguy';
		$repo	= 'WP-Plugin-Directories';
		new wp_github_updater( array(
			 'slug'               => plugin_basename( __FILE__ )
			,'proper_folder_name' => dirname( plugin_basename(__FILE__) )
			,'api_url'            => "{$http}api.{$host}/repos/{$name}/{$repo}"
			,'raw_url'            => "{$http}raw.{$host}/{$name}/{$repo}/master"
			,'github_url'         => "{$http}{$host}/{$name}/{$repo}"
			,'zip_url'            => "{$http}{$host}/{$name}/{$repo}/zipball/master"
			,'sslverify'          => true
			,'requires'           => $wp_version
			,'tested'             => $wp_version
			,'readme_file'        => 'readme.md'
			,'description'        => array( 
				'changelog' => $this->update_message()
			 )
		) );
	}


	/**
	 * Callback to set the SSL verification for HTTP requests to GitHub to false
	 * 
	 * @since  1.0
	 * @param  array  $args
	 * @param  string $url
	 * @return array  $args
	 */
	public function update_request_args( $args, $url )
	{
		// Only needed once - this saves us checking the $url
		remove_filter( current_filter(), __FUNCTION__ );

		return array_merge( $args, array(
			'sslverify' => false
		) );
	}


	/**
	 * Displays an update message for plugin list screens.
	 * Shows only the version updates from the current until the newest version
	 * 
	 * @uses WordPress HTTP API
	 * @uses $allowed_html and $allowed_protocolls inside `install_plugin_information()`
	 * 
	 * @since  0.9
	 * @return string The actual Output message
	 */
	public function update_message()
	{
		// Get `changelog.txt` from GitHub via WP HTTP API
		$remote_data = wp_remote_get( 
			 $this->remote_changelog
			,false 
		);

		// Die silently
		$response = wp_remote_retrieve_response_code( $remote_data );

		if ( is_wp_error( $remote_data ) )
			return _e( 'No changelog could get fetched.', 'cd_apd_textdomain' );

		if ( 404 === $response )
			return $remote_data['response']['message']; 

		return sprintf( 
			 "<p style='font-weight:normal;'>%s</p>"
			,wp_remote_retrieve_body( $remote_data )
		);
	}
} // END Class CD_APD_Bootstrap

} // endif;

example_plugin/register_addt_plugins_dir.php
<?php
! defined( 'ABSPATH' ) AND exit();
/*
Plugin Name:  Additional Plugin Directories: Example Plugin
Plugin URI:   http://github.com/chrisguitarguy
Description:  Example Plugin to show how to register additional plugin directories
Version:      0.1
Author:       Franz Josef Kaiser
Author URI:   http://unserkaiser.com
License:      MIT
*/


/**
 * Registers a new plugin directory
 * 
 * @example
 * $args Array (Valid args for `root`) 'content', 'plugins', 'muplugins', 'root'
 * The new directories must be subdirectories of the following WP file system constants:
 * 'content':   (default) WP_CONTENT_DIR
 * 'plugins':   WP_PLUGIN_DIR
 * 'muplugins': WPMU_PLUGIN_DIR
 * 'root':      one level below WP_CONTENT_DIR
 * 
 * @return void
 */
function cd_apd_register_additional_plugin_directories()
{
	// Better abort - if we don't do this, we'll create an error on deactivation of the main plugin.
	if ( ! function_exists( 'register_plugin_directory' ) )
		return;

	// Call the public API function once for every directory you want to add.
	register_plugin_directory( array( 
		 'dir'   => 'example_plugin_directory'
		,'label' => 'Example Label for the list table'
		,'root'  => 'plugins'
	) );
}
// Needs to be added on the `plugins_loaded` hook with a priority of `0`.
add_action( 'plugins_loaded', 'cd_apd_register_additional_plugin_directories', 0 );


inc/api.php
<?php
! defined( 'ABSPATH' ) AND exit();



/**
 * The central API
 * 
 * @author     Christopher Davis, Franz Josef Kaiser
 * @license    GPL
 * @copyright  © Christopher Davis, Franz Josef Kaiser 2011-2012
 * 
 * @since      0.3
 * @package    Additional Plugin directories
 * @subpackage The central API
 */


/**
 * Registers a new plugin directory.
 * 
 * @since  0.1
 * @uses   _get_new_plugin_directory_root()
 * @param  array  $args             An Array of arguments: 'dir' = Name of the directory, 'label' = What you read above the list table, 'case' = Where the dir resides.  
 * @param  string $deprecated_dir   (deprecated arg) The new plugin directory. Either a full path or a folder name within wp-content.
 * @param  string $deprecated_label (deprecated arg) The nice name of the plugin directory. Presented in the list table.
 * @return bool                     TRUE on success, FALSE in case the $key/$label is already in use.
 */
function register_plugin_directory( $args, $deprecated_dir = '', $deprecated_label = '' )
{
	// The call was too late (or too early in case of a MU-Plugin)
	if ( 'plugins_loaded' !== current_filter() )
	{
		_doing_it_wrong( 
			 __FUNCTION__
			,__( 'Registering a new plugin directory should be done during the `plugins_loaded` hook on priority `0`.', 'cd_apd_textdomain' )
			,'0.1'
		);
	}

	// Deprecating single arguments
	if ( ! is_array( $args ) )
	{
		$info = debug_backtrace();
		// The key now gets built from the label: Converted to lowercase alphanumeric string.
		_deprecated_argument( 
			 __FUNCTION__
			,'0.3'
			,sprintf( 
				 __( "%sYou need to specify the arguments &ndash; when registering new plugin directories &ndash; as associative array.%s%s", 'cd_apd_textdomain' )
				,'<br /><blockquote><strong>'
				,'</strong></blockquote>'
				,"The call was from within: <code>{$info[ 0 ]['file']}</code> in the function: <code>{$info[ 1 ]['function']}()</code>.<br />"
			)
		);

		// Fix for back compat
		$args = array(
			 'dir'   => $deprecated_dir
			,'label' => $deprecated_label
		);
	}

	// Setup defaults
	$args = wp_parse_args( $args, array( 'root' => 'content' ) );

	global $wp_plugin_directories;

	empty( $wp_plugin_directories ) AND $wp_plugin_directories = array();

	$new_dir = _get_new_plugin_directory_root( $args['root'] ).$args['dir'];

	if ( ! file_exists( $args['dir'] ) AND file_exists( $new_dir ) )
	{
		$args['dir'] = $new_dir;
	}

	// Build $key from $label
	$key = strtolower( preg_replace( "/[^a-zA-Z0-9\s]/", "", $args['label'] ) );

	// Return FALSE in case we already got the key
	if ( isset( $wp_plugin_directories[ $key ] ) )
		return false;

	// Assign the directory
	$wp_plugin_directories[ $key ] = array(
		 'dir'   => $args['dir']
		,'label' => $args['label']
		,'root'  => $args['root']
	);

	return true;
}


/**
 * Retrieves the root path for the new plugin directory.
 * 
 * @internal Callback function for register_plugin_directory()
 * 
 * @since    0.3
 * @param    string $case Valid: 'content', 'plugins', 'muplugins', 'root'.
 * @return   string $root The root path based on the WP filesystem constants.
 */
function _get_new_plugin_directory_root( $root )
{
	switch ( $root )
	{
		case 'plugins' :
			$root = WP_PLUGIN_DIR;
			break;

		case 'muplugins' :
			$root = WPMU_PLUGIN_DIR;
			break;

		// Experimental Edge Case: 
		// Assuming that the WP_CONTENT_DIR is a direct child of the root directory
		// and directory separators are "/" above that. 
		// Maybe needs enchancements later on. Wait for feedback in Issues.
		case 'root' :
			$root = explode( DIRECTORY_SEPARATOR, WP_CONTENT_DIR );
			$root = explode( '/', array_pop( $root ) );
			$root = array_pop( $root );
			break;

		case 'content' :
			$root = WP_CONTENT_DIR;
			break;

		default :
			$root = apply_filters( "adp_root_{$root}", WP_CONTENT_DIR );
			break;
	}

	return trailingslashit( $root );
}
